### PR TITLE
[NET-9201] Validating webhook for registrations

### PR DESCRIPTION
**Pull request by jm96441n**
### Changes proposed in this PR ###  
- add validating webhook to catch configuration errors earlier

### How I've tested this PR ###
- requirements: `kubectl` and `yq` to be installed on your system, this branch checked out in consul-k8s
- set the following values in your shell, update the k8s chart location value to be the path to the consul-k8s charts on your machine
```
export CONSUL_K8S_CHARTS_LOCATION="$HOME/hashi/consul-k8s/charts/consul"
export CONSUL_HTTP_ADDR="127.0.0.1:8501"
export CONSUL_HTTP_SSL=true
export CONSUL_HTTP_SSL_VERIFY=false
```
- build from this branch `make docker-dev`
- from the following gist run the `start.sh` file, this will set up your cluster and register the service: https://gist.github.com/jm96441n/43e526387fc83f7079a10a555f739177
- open up the consul ui at `https://localhost:8501` (you'll have to allow insecure traffic over https here)
- see the `zoidberg` service registered
- make some modification to the `zoidberg` registration file at `registration.yaml` such as:
    1. remove the `node`, `address`, `service.name`, `healthCheck.Definition.IntervalDuration`,  and/or `healthCheck.CheckID` fields
    2. change any of the durations to an invalid duration
 - run `kubectl apply -f registration.yaml` and see the command error out

### How I expect reviewers to test this PR ###
run the above
read the code
run the tests

### Checklist ###
- [X] Tests added
- [ ] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -340,7 +340,14 @@ rebase the branch on main, fixing any conflicts along the way before the code ca
 1. Replace the names
 1. Ensure you've correctly replaced the names in the kubebuilder annotation, ensure the plurality is correct
     ```go
-    // +kubebuilder:webhook:verbs=create;update,path=/mutate-v1alpha1-ingressgateway,mutating=true,failurePolicy=fail,groups=consul.hashicorp.com,resources=ingressgateways,versions=v1alpha1,name=mutate-ingressgateway.consul.hashicorp.com,webhookVersions=v1beta1,sideEffects=None
+      // +kubebuilder:webhook:verbs=create;update,path=/mutate-v1alpha1-ingressgateway,mutating=true,failurePolicy=fail,groups=consul.hashicorp.com,resources=ingressgateways,versions=v1alpha1,name=mutate-ingressgateway.consul.hashicorp.com,sideEffects=None,admissionReviewVersions=v1beta1;v1
+    ```
+1. Ensure you update the path to match the annoataing in the `SetupWithManager` method:
```
Comment (NiniOak) [inline CONTRIBUTING.md:345]: typo: annotation

File: control-plane/api/v1alpha1/registration_webhook.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+// +kubebuilder:webhook:verbs=create;update,path=/validate-v1alpha1-registration,mutating=false,failurePolicy=fail,groups=consul.hashicorp.com,resources=registrations,versions=v1alpha1,name=validate-registration.consul.hashicorp.com,sideEffects=None,admissionReviewVersions=v1beta1;v1
+
+func (v *RegistrationWebhook) Handle(ctx context.Context, req admission.Request) admission.Response {
+	resource := &Registration{}
+	decodeErr := v.decoder.Decode(req, resource)
+	if decodeErr != nil {
+		return admission.Errored(http.StatusBadRequest, decodeErr)
+	}
+
+	v.Logger.Info("validating registration", "name", resource.Name, "timeout", resource.Spec.HealthCheck.Definition.TimeoutDuration)
```
Comment (NiniOak) [inline control-plane/api/v1alpha1/registration_webhook.go:42]: ~ Seeing as this HealthCheck field `TimeoutDuration string` is optional (not sure if we have a default timeout value), should it be added to the logger?

File: control-plane/api/v1alpha1/registration_webhook_test.go [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+func TestValidateRegistration(t *testing.T) {
+	cases := map[string]struct {
+		newResource   *Registration
+		expAllow      bool
```
Comment (NiniOak) [inline control-plane/api/v1alpha1/registration_webhook_test.go:23]: what does exp in expAllow mean? expect?

File: control-plane/api/v1alpha1/registration_webhook_test.go [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+						Name:   "check name",
+						Status: "passing",
+						Definition: HealthCheckDefinition{
+							IntervalDuration: "150",
+						},
+					},
+				},
+			},
+			expAllow:      false,
+			expErrMessage: "invalid registration.Spec.HealthCheck.Definition.IntervalDuration value: \"150\"",
```
Comment (NiniOak) [inline control-plane/api/v1alpha1/registration_webhook_test.go:206]: I'm wondering if this should be a more descriptive error message. Are we expecting int only for duration in seconds?

File: control-plane/api/v1alpha1/registration_webhook_test.go [hunk 5]
```diff
@@ -0,0 +1,330 @@
...
+						Definition: HealthCheckDefinition{
+							IntervalDuration: "10s",
+						},
+					},
+				},
+			},
+			expAllow:      false,
+			expErrMessage: "invalid registration.Spec.HealthCheck.Status value, must be 'passing', 'warning', or 'critical', actual: \"wrong\"",
+		},
+		"everything that can go wrong has gone wrong": {
```
Comment (NiniOak) [inline control-plane/api/v1alpha1/registration_webhook_test.go:274]: 😆

File: control-plane/api/v1alpha1/registration_webhook.go (hunk 2 above)
Comment (jm96441n) [reply control-plane/api/v1alpha1/registration_webhook.go:42]: oh whoops, this was a debugging statement

File: control-plane/api/v1alpha1/registration_webhook_test.go (hunk 3 above)
Comment (jm96441n) [reply control-plane/api/v1alpha1/registration_webhook_test.go:23]: `expectedAllow` I can expand this (I copied/pasted from another webhook test)

File: control-plane/api/v1alpha1/registration_webhook_test.go (hunk 4 above)
Comment (jm96441n) [reply control-plane/api/v1alpha1/registration_webhook_test.go:206]: so it needs to be something like `5m` or `30s` (from our docs https://developer.hashicorp.com/consul/docs/services/configuration/checks-configuration-reference)

**Review by NiniOak — APPROVED** [review]
Validation could be quicker but looks good. Thanks for adding this

**Pull request merged**